### ISSUE TITLE
import.meta.main

### ISSUE THREAD

**Issue by MylesBorins** [nodejs/modules#274]
Deno just added this.

Should we?

https://github.com/denoland/deno/pull/1835

**Comment by ljharb** [comment]
I'm a bit confused what it does - does it mean "it was the `main` in package.json"? it was the primary entry point? what about `import()`?

**Comment by devsnek** [comment]
it's true if the file was the entrypoint. for node i'd rather it be called something else cuz the term main already has package-scoped meaning.

**Comment by ljharb** [comment]
Right, but anything that's `import()`ed is also an entry point, and I'm very unclear on why you care what that value is. Currently you can look at `process.argv` and `__filename` to determine that (and `import.meta.filename` or similar is quite well motivated).

Also yes, `main` is a terrible name for multiple reasons, not the least of which is that it's a package.json field.

**Comment by guybedford** [comment]
I really like the `import.meta.main` concept. Currently in Node.js it is common for CLI tools to do the following check:

```js
if (require.main === module) {
  console.log('This is a CLI tool! Usage: ...');
}
```

We currently have no version of the above check in ECMAScript modules, so this is a problem that comes up, and there is currently no easy way to do this without it feeling like a hack.

As an already-paved use case, users are familiar with the `main` terminology in this context already.

The above check then becomes:

```js
if (import.meta.main) {
  console.log('This is a CLI tool! Usage: ...');
}
```

and the really nice thing about this pattern is that it works in browsers and other environments completely fine without any need for compatibility API layers or compilation.

For these reasons I'm 100% supportive of `import.meta.main`.

**Comment by ljharb** [comment]
Doesn’t that use case predate the current community preference, which is to have separate packages for a library and a CLI?

**Comment by guybedford** [comment]
@ljharb this use case is explicitly documented in https://github.com/nodejs/node/blob/master/doc/api/modules.md#accessing-the-main-module.

**Comment by ljharb** [comment]
Sure but so is every aspect of the require algorithm :-) the better question is, is it still a common or desired use case to determine if something is the entry point, and how does that change now that dynamic import allows multiple entry points?

**Comment by guybedford** [comment]
@ljharb yes it is a desired feature. The `require.main === module` was something we had to explicitly support in the ncc project as users were using it in CLI tools. This came up pretty early in the project and had multiple users asking about it here - https://github.com/zeit/ncc/issues/224. Dynamic import does not affect this, as what we are distinguishing is the CLI entry point, not the module graph entry point.

**Comment by ljharb** [comment]
The feature seems fine then (it’d only be true in the top-level process entry point); the name, definitely not (but we can bikeshed that).

**Comment by bmeck** [comment]
Does this absolutely need to be in our first iteration or can we get feedback / focus on other things?

**Comment by zenparsing** [comment]
Thanks for bringing this up @MylesBorins. I (think) we can say this is post-MVP but it's good to have on our radar.

**Comment by MylesBorins** [comment]
can definitely be post MVP. Just wanted to document what's going on in ecosystem

**Comment by shian15810** [comment]
Well, I'm not part of the organization, but this proposal seems so interesting.

Can I try implementing this in https://github.com/nodejs/node and possibly a PR?

**Comment by devsnek** [comment]
I'm not a huge fan of the concept of main. imo you should have separate files for bin and lib, which is something other languages do just fine, and installers like npm and yarn already let you specify separate bins.

**Comment by guybedford** [comment]
@shian15810 thanks for showing an interest in the modules work! Contributions are very much welcome to Node.js core and the module work here.

We do have some consensus issues for this feature due to our not wanting to provide unnecessary features without seeing a strong need for them in the initial modules implementation. `import.meta` is a very widely used namespace (every module!) with strong backwards compatibility needs, so we do also need to be very cautious about what we put on it.

A PR to Node.js core would certainly drive discussion, and may even sway consensus. Also hearing more about your use case as a CLI maintainer and how this is useful could help us better understand the importance of the feature to you (including for example the points raised by Gus above). Even if the PR sits without approval, we may be able to come back to it in a few months even as well.

**Comment by shian15810** [comment]
The use case of mine is that both `bin` and `main` fields in package.json point to the same file. Using `require.main === module`, as mentioned in the official docs [here](https://nodejs.org/api/modules.html#modules_accessing_the_main_module), to tell whether the script has been run directly or not, so that to know if the main function should be executed or exported.

As far as I can tell, with a package.json having `"type": "commonjs"`, there are three ways to tell if a file is run directly (aka not being `require`d):

1. `require.main === module` or `process.mainModule === module`
2. `require.main.filename === __filename` or `process.mainModule.filename === __filename`
3. `!module.parent`

However, with a package.json having `"type": "module"`, `require` and `module` are not available, as well as `process.mainModule` is `undefined`.

In addition, according to [this](https://github.com/nodejs/node/issues/15760#issuecomment-400799290), `module.parent` in a cjs script will be `undefined` if the parent is a mjs script. After some testing, `require.main` and `process.mainModule` are both `undefined` too in this context.

What's left is just `import.meta.url`, but the problem is what to compare with to know if it is the main script.

In my imagination, if `import.meta.main` is implemented, I could use it in mjs script, while retaining `require.main === module` in cjs script. But this will still pose a serious problem to a project with mixed cjs and mjs scripts importing each other.

So yeah, I agree with @devsnek in this case, that is pointing `bin` and `main` to separate scripts in the first place, which I now think is the right way to this problem.

As a last note, fI've read somewhere in this repo stating that in the context of esmodule, there is no concept of main like in commonjs, correct me if I'm wrong. But I certainly think that there must be some other use cases of determining the main script, just like `require.main === module` in commonjs. It is **good to have** `import.meta.main` as the counterpart in esmodule.

**Comment by MylesBorins** [comment]
Going to abandon this for now, please feel free to reopen

**Comment by guybedford** [comment]
I still believe this is an important feature, to be put in the same basket as _CJS features not available in ESM_ like `require.resolve`.

It's not a common feature, but it is definitely a heavily ingrained one where it is used, and has a small collection of users that absolutely do expect this functionality (mainly CLI developers).

**Comment by devsnek** [comment]
i am still against the concept of a main module in general.

**Comment by SMotaal** [comment]
@guybedford Which labels are relevant here? Can you add them to make it easier to related to the context please — ie `cjs` anr/or `esm`... etc.

**Comment by SMotaal** [comment]
@devsnek Can you elaborate more? Main here is the main entrypoint, so it is like window.location where there is a browser context or self.location where there is not one.

**Comment by SMotaal** [comment]
@guybedford Awesome, that really helps 💯

**Comment by devsnek** [comment]
@SMotaal i dislike the idea of differentiating entrypoints. you can just use a separate file for your bin.

**Comment by devsnek** [comment]
being an entrypoint doesn't inherently mean it's a cli (for example, cf workers). additionally, some hosts like browsers don't map well to the concept of entrypoints. depending on how you think about it, a webpage may either have several or no entrypoints, and you choose to think about it having several, none of them would be a cli.

**Comment by SMotaal** [comment]
@devsnek certainly it is one style, I am not saying your style is wrong, but others also have justifications that cannot be open to judgement because that is comes down to matters of opinion.

As it happens, any given module in a package can be called as a result of at least two paths, a main entrypoint in the package, or an entrypoint for which the package is a direct or indirect dependency… some packages may actually consider this a functional parameter depending on their purpose.

So no disagreement or disapproval is intended, just wanting us to consider the more diverse landscapes that we may not ourselves be interested in (yet).

**Comment by guybedford** [comment]
To clarify some of the above, the only module that gets `import.meta.main` set is the module passed into `node main.js` it is only the module corresponding to `process.argv[1]`. And no, it doesn't apply to browsers.

This is the Node.js application entry point main, not package entry points, or realm / worker entry points.

An equivalent way to achieve this check would be to do `path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)`.

Perhaps the above explicit check would be enough though.

**Comment by guybedford** [comment]
To clarify some of the above, the only module that gets `import.meta.main` set is the module passed into `node main.js` it is only the module corresponding to `process.argv[1]`. And no, it doesn't apply to browsers.

This is the Node.js application entry point main, not package entry points, or realm / worker entry points.

An equivalent way to achieve this check would be to do `path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)`.

Perhaps the above explicit check would be enough though.

**Comment by devsnek** [comment]
@guybedford what i'm trying to say is that because it depends on how you approach it, we shouldn't try to reify a certain pattern. if the condition for your cli is that it is the argument to node, that check seems good to me. i agree it's not the most ergonomic, but i think that comes down to the awkardness of paths and urls.

**Comment by SMotaal** [comment]
@guybedford thanks for the reminder/clarification… so a consideration that comes to mind, who has the authority to know.

And so it being a "readonly" and "private" field on `import.meta` has very different implications than globals, where it is possible to run into all kinds of costs to try to check for distortions.

Just a thought, no strong feelings on it personally.

**Comment by guybedford** [comment]
The existing Node.js pattern it is replacing is the `require.main === module` pattern which is documented in the Node.js docs for require.

**Comment by devsnek** [comment]
the pattern for cjs, sure. i'd argue there is no pattern for esm yet. i'd like to lean on the more mature advancements in the ecosystem like package manager bin entries instead of continuing to copy old python patterns.

**Comment by guybedford** [comment]
The point for keeping this open is only that this falls under _things supported in CJS but not ESM_ right now and that there are many users who would expect this functionality. That doesn't mean we have to do it, but only that we shouldn't neglect those existing user expectations.

**Comment by devsnek** [comment]
right, i'm saying we should put a section in the docs saying "for those of you used to including your lib exports and bin logic in the same file, the esm pattern is to use two separate files". i'd be happy to write it up too if this discussion finishes in favor of that.

**Comment by addaleax** [comment]
@devsnek I find it curious that you’re both saying “we shouldn't try to reify a certain pattern” and “the esm pattern is to use two separate files”.

I heavily agree with the first sentiment, which is why I feel we should *not* enforce using two separate files rather than a single one.

**Comment by bmeck** [comment]
I don't think we are explicitly forcing using two files as we have pointed out a way of doing this, and other workarounds do exist. I think the using of two files is just a natural feeling solution compared to other workarounds and is why people think of it as a specific pattern to use. I personally remain unsure of exposing a new property on `import.meta` to fill this feature; workarounds already exist so perhaps we can lean on those until the need to expose something proves to be urgent.

**Comment by jkrems** [comment]
There's also the issue with block-scoped loading of dependencies. In CJS `require.main` was nice because it allowed inlining things like:

```js
if (require.main === module) {
  const assert = require('assert');
  const http = require('http');

  // Setup example and/or test
}
```

In ESM, that pattern either needs to use an awkward `await import('assert')` or, more likely, importing those test/example dependencies at the top of the file. I wouldn't be too eager to add `import.meta.main` unless we see people use one of the workarounds described above in "real" code. If `isMain(import.meta)` ends up a popular npm package, we can always add the property.

**Comment by devsnek** [comment]
@addaleax good point :) To clarify: I was approaching it from the perspective of what people would do if node doesn't add anything special for this case. Like guy said, there are more patterns, like checking argv. Maybe it would be best not to document anything specifically though.

**Comment by azz** [comment]
Just got blocked on this when switching some apps to use native ESM. We have the same, very common pattern of:

```js
function main() {}

if (require.main === module) {
  main();
}
```

Replacing this with `if (import.meta.main) {}` would be very desirable. It's actually an improvement on the CJS version.

**Comment by devsnek** [comment]
@azz is using separate files not acceptable for some reason? more info would be good.

**Comment by azz** [comment]
Moving to a separate file is fine in some cases, but can cause some tricky refactoring if there are dependants on the existing behaviour. It can be a breaking change.

**Comment by azz** [comment]
Another use case is scripts/CLIs that are distributed as a single JS file.

**Comment by tschaub** [comment]
This would be a pretty nice upgrade path for people wanting to use modules with Node:
```diff
--- a/nice.js
+++ b/nice.js
@@ -1,4 +1,4 @@
 
-if (require.main === module) {
+if (import.meta.main) {
   main();
 }
```

I'm glad to know about this workaround, but this is more awkward in my opinion:
```diff
--- a/not-as-nice.js
+++ b/not-as-nice.js
@@ -1,4 +1,6 @@
+import {fileURLToPath} from 'url';
+import process from 'process'
 
-if (require.main === module) {
+if (process.argv[1] === fileURLToPath(import.meta.url)) {
   main();
 }
```

It's not hard to find lots (55K+) of examples following this pattern: https://github.com/search?l=&q=%22require.main+%3D%3D%3D+module%22+language%3AJavaScript&type=Code

**Comment by tschaub** [comment]
A shortcoming of the `process.argv[1] === fileURLToPath(import.meta.url)` workaround is that it doesn't work if someone runs the script without the `.js` extension.

For example, the workaround works when invoked like this:
```bash
# here the script can detect it was run directly
node script.js
```

But not when invoked like this:
```bash
# here the script can't detect it was run directly
node script
```

**Comment by jkrems** [comment]
One crazy idea: what if node invoked an exported function main instead? If the entry point exports a function called main, it is invoked after the module graph is initialized. Maybe even passing argv..? That seems like a much more testable (and clean) pattern than top level conditional code.

**Comment by ljharb** [comment]
Ugh, we're not C++ or java, let's not do that :-/

**Comment by jkrems** [comment]
It’s not just Java/C++ that does it. I’m not sure if there’s a more complete argument than “languages I don’t like do something that looks similar”..?

**Comment by devsnek** [comment]
`module === require.main` is a fine pattern for commonjs, but esm is not commonjs. I don't think we need to feel forced to find a solution. It should be expected that migrating to a new system involves making changes that are more than a single line of code.

**Comment by addaleax** [comment]
@jkrems How about “It’s easier to make the very first steps with JS/Node.js this way”? I know that might seem like it’s not much in the big picture, but I do like this about JS.

**Comment by bmeck** [comment]
[WASI is going to be taking the main() route](https://github.com/WebAssembly/interface-types/blob/linking/proposals/interface-types/linking/Explainer.md#command-modules-revisited) (different from [start](https://webassembly.github.io/spec/core/syntax/modules.html#syntax-start)); this might be something to discuss.

**Comment by jkrems** [comment]
Just to be clear: I wasn’t suggesting removing top level code. I was suggesting this as a way to mark some top level code as conditionally executed only when the file is running as the entrypoint. I hope that in most cases (especially for beginners) that’s not something people would have to worry about. :) Having separate files is almost always easier.

But yes, I can see why to some people a special condition + if block could be easier to explain than a “magical” function name. Just tried to understand if that’s what Jordan had in mind or if there are other arguments we should consider.

**Comment by devsnek** [comment]
@bmeck probably something to discuss on the esm proposal instead? We'd want all the different js environments that support wasi to do it uniformly.

**Comment by bmeck** [comment]
@devsnek for WASI, it is just the `"main"` string.

**Comment by devsnek** [comment]
@bmeck right, i mean the semantics of handling that main.

**Comment by bmeck** [comment]
We might also want to lock step with WASM interface-types so that WASM side doesn't go out of sync with ESM

**Comment by ljharb** [comment]
A complication is that there's more questions that may potentially need answering than "is this file the entry point" - there's also "is this file a loader entry point", "is this file being consumed by a loader entry point", "is this file used in --require", "is this file being consumed within --require", "was this file eval'd", etc. `require.main` doesn't answer all these questions either :-/

**Comment by bmeck** [comment]
@ljharb I'd agree that question isn't solved here; if there is a more universal solution we might want to look at other things. I think adding conditions for things like `--require` style usage would be nice but likely is a different topic and might be better served with a different mechanism than how an entry point is bootstrapped. Even if we call `main()` it doesn't mean that the file is the only entrypoint, and userland could always call `main()` on other modules manually.

**Comment by tschaub** [comment]
In case others find it useful, [`es-main`](https://www.npmjs.com/package/es-main) is a package that allows for a check similar to `require.main === module`.
```js
import esMain from 'es-main';

if (esMain(import.meta)) {
  // Do something special.
}
```

This works for `node script.js`, `node script`, and `./script.js` type use.

**Comment by devsnek** [comment]
@tschaub that would appear to have some bugs involving process.argv[1] not being present (repl) and returning true when i assume you would want it to return false (argv[1] is `a`, `a.x` is loaded, you call that from `a.y` and it returns true).

**Comment by tschaub** [comment]
@devsnek - thanks for the report.  Pull requests accepted :)

**Comment by aduh95** [comment]
I have opened a PR to implement the feature on node repo, does anyone want to raise their voice regarding the value you would expect `import.meta.main` to take on a `Worker` thread?

From https://github.com/nodejs/node/pull/32223#issuecomment-598355387
> > What would the value be in a Worker thread?
> 
> Good question! With the current implementation, it is `true` for the Worker entry point, and `false` for sub-modules. To check if the current script is run as CLI, that means you also have to check for `isMainThread` value... I'm not sure if that's the behaviour we want to implement, but it definitely needs documentation of this topic.

FWIW Deno follows the same logic.

**Comment by trusktr** [comment]
If something like `es-main` exists (and can be improved for edge cases), maybe we can close this? It's working great for me in replacement of `require.main === module`.

**Comment by kaizhu256** [comment]
+1
as contributor to [jslint](https://github.com/jslint-org/jslint) this feature would be useful.  latest version of jslint doesn't use npm for installation.

ideally, i would like jslint to be a tool-less, self-contained file able to run both from cli/esm.

```javascript
#!/bin/sh

# install tool-less jslint.mjs by simply downloading self-contained file
curl https://www.jslint.com/jslint.js > jslint.mjs

# run jslint.mjs from cli
node jslint.mjs foo.js // jslint foo.js

# use jslint.mjs as imported esm
node --input-type=module -e "////'"'
import jslint from "./jslint.mjs"
let code = "..."
let result = jslint(code);
result.warnings.forEach(function (warning) {
    console.error(warning);
});
'
```

**Comment by loynoir** [comment]
FYI, there is a, similar but not same, magic comment  `/*@__PURE__*/` in bundler, Supported by rollup, esbuild, etc.

Which allow you to remove command line only code from output.
```js
// library code

/* @__PURE__ */(async () => {
    // command line only code
})();
```

**Comment by loynoir** [comment]
```js
// library code

/*@__PURE__*/(async () => {
    // command line only code

    if ((await import('es-main')).default(import.meta)) {
        // command line only main code
    }
})()
```

**Comment by BlueNebulaDev** [comment]
I'm having the impression that this feature is not getting implemented because some node developers "dislike the idea of differentiating entrypoints" and want to force us to "use a separate file for your bin".

Clearly A LOT of users want this feature and this feature costs NOTHING to node. Do we really need to complicate our lives to use `import.meta` because some power tripping mod wants to dictate how we are supposed to code?!

Can we pick any solution and go with it FFS?

**Comment by Beedeebee** [comment]
I read all the comments but I can't understand why this issue is still open after 3 years but it's not going anywhere.

Are you still stuck deciding whether you want this feature at all, or what else is blocking this? After longer than three years and a hundred comments I think it's time to decide: please close the issue if you really don't want it.

Otherwise, if you are still really undecided, why don't you add something like a `import.meta.nodeJsExperimentalMain` property?

- The name alone makes it very clear that this feature is likely to disappear at any point. Good code won't depend on it
- It would be a way to experiment with it. It could help us understand its benefits and problems and would help you decide whether you want this feature and how to implement it
- The users who don't mind its experimental status could use this feature

---

This is why I would love to have this feature, even experimentally.

When I quickly jot down some prototype, I like to include in some modules something to quickly test them. A bunch of asserts and printing out some values. It's code that I'm writing quickly to see how a certain library or design feels like. Most of the times I end up discarding this code, or heavily refactoring it: the module's API is almost surely guaranteed to change.

Having some dedicated unit tests or extra files for this stuff feels way too much, to the point that I'm using CommonJS instead of ES modules only because of this... I would love to use `import.meta.nodeJsExperimentalMain` for this. Then, once my code is a bit more mature and I start refactoring it, I'd set up mocha and move this testing code into proper tests.

**Comment by rmclaughlin-nelnet** [comment]
@JakobJingleheimer Sorry I should have been more clear. We are trying to upgrade this module to ESM. So the require object is not available. My question is how do we duplicate this functionality without the new feature mentioned above.

**Comment by brianjenkins94** [comment]
https://stackoverflow.com/questions/57838022/detect-whether-es-module-is-run-from-command-line-in-node

**Comment by rmclaughlin-nelnet** [comment]
> https://stackoverflow.com/questions/57838022/detect-whether-es-module-is-run-from-command-line-in-node
> 
> ```ts
> import url from "url";
> 
> if (import.meta.url === url.pathToFileURL(process.argv[1]).href) {
>   // module was not imported but called directly
> }
> ```

Unfortunately that wont work for us because we call our scripts without the extension
https://github.com/nodejs/modules/issues/274#issuecomment-588454524

**Comment by tschaub** [comment]
The [`es-main`](https://www.npmjs.com/package/es-main) package accounts for this and other nuances.  See https://github.com/nodejs/modules/issues/274#issuecomment-588528200

**Comment by bricker** [comment]
`require.module === main` is particularly useful when writing and testing Node GitHub Actions, although I could make this same argument for any CLI. The entrypoint file for a Node action is defined in the `action.yml` file, and when that file is executed it must perform the action. A file that only calls another function is cumbersome, especially when you have dozens of these actions.

Some examples:

### Using `require.module === main`

entrypoint.js

```js
const core = require('@actions/core');
module.exports = function run(deps) { }

if (require.main === module) {
  run({ core });
}
```

entrypoint.test.js

```js
const test = require('node:test');
const assert = require('node:assert');
const sinon = require('sinon');
const run = require('./entrypoint');

test('run', (t) => {
  run({ core: sinon.stub() });
  assert(true);
});
```

### Using a separate "cli" file

run.js

```js
module.exports = function run(deps) { }
```

entrypoint.js

```js
const core = require('@actions/core');
const run = require('./run');
run({ core });
```

entrypoint.test.js

```js
const test = require('node:test');
const assert = require('node:assert');
const sinon = require('sinon');
const run = require('./run');

test('run', (t) => {
  run({ core: sinon.stub() });
  assert(true);
});
```

I much prefer the `require.main === module` solution, for the following reasons:

1. One fewer file to maintain.
2. cli `entrypoint.js` files are repetitive and often identical.
3. imports of dependencies are in a different file from the function that receives them. This requires a developer to update two files to add or remove a single dependency, and results in unused dependencies still being imported and passed through.

**Comment by sosoba** [comment]
Simple loader workaround:

entrypoint.js
```js
if (import.meta.main ) {
  // start
}
```

mainLoader.js
```js
import {pathToFileURL} from 'node:url';
import {argv} from 'node:process';
import {Buffer} from 'node:buffer';

const mainUrl = pathToFileURL(argv[1]).href;

export const load = async (specifier, context, nextLoad) => {
  if (specifier === mainUrl) {
    let {source, ...rest} = await nextLoad(specifier, context, nextLoad);
    if (Buffer.isBuffer(source)) {
      source = Buffer.concat([Buffer.from('import.meta.main = true;'),source]);
      return {source, ...rest};
    }
  }
  return nextLoad(specifier, context, nextLoad);
};
```

```sh
node --loader=mainLoader.js entrypoint.js
```

**Comment by Gaubee** [comment]
I use `import.meta.url.endsWith(process.argv[1].replace(/\\/g,'/'))` work in node 18+

**Comment by lwr** [comment]
https://github.com/nodejs/modules/issues/274#issuecomment-588528200
> This works for `node script.js`, `node script`, and `./script.js` type use.

Unfortunately if we are having `bin/index.js` this doesn't work for `node bin` but only `node bin/index.js`, `node bin/index`, and `./bin/index.js`.

so we have to change the detecting method to
```js
import {createRequire} from 'module';
import {fileURLToPath} from 'url';

if (process.argv[1] && createRequire(process.argv[1]).resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
    // main();
}

// or
const esMain = meta => ((process.argv[1] && createRequire(process.argv[1]).resolve(process.argv[1])) === fileURLToPath(meta?.url));
if (esMain(import.meta)) {
    // main();
}
```
